components/membership/locally-hosted-identities-service-impl/src/main/kotlin/net/corda/membership/locally/hosted/identities/impl/LocallyHostedIdentitiesServiceImpl.kt

```
@@ -0,0 +1,214 @@
+import net.corda.configuration.read.ConfigurationReadService
+import net.corda.data.p2p.HostedIdentityEntry
+import net.corda.libs.configuration.helper.getConfig
+import net.corda.lifecycle.LifecycleCoordinatorFactory
+import net.corda.lifecycle.LifecycleCoordinatorName
+import net.corda.lifecycle.LifecycleEvent
+import net.corda.lifecycle.LifecycleStatus
+import net.corda.lifecycle.RegistrationStatusChangeEvent
+import net.corda.lifecycle.StartEvent
+import net.corda.lifecycle.StopEvent
+import net.corda.lifecycle.createCoordinator
+import net.corda.membership.locally.hosted.identities.IdentityInfo
+import net.corda.membership.locally.hosted.identities.LocallyHostedIdentitiesService
+import net.corda.messaging.api.processor.CompactedProcessor
+import net.corda.messaging.api.records.Record
+import net.corda.messaging.api.subscription.config.SubscriptionConfig
+import net.corda.messaging.api.subscription.factory.SubscriptionFactory
+import net.corda.schema.Schemas
+import net.corda.schema.configuration.ConfigKeys
+import net.corda.v5.base.exceptions.CordaRuntimeException
+import net.corda.v5.base.util.contextLogger
+import net.corda.v5.base.util.seconds
+import net.corda.virtualnode.HoldingIdentity
+import net.corda.virtualnode.toCorda
+import org.osgi.service.component.annotations.Activate
+import org.osgi.service.component.annotations.Component
+import org.osgi.service.component.annotations.Reference
+import java.security.cert.CertificateFactory
+import java.security.cert.X509Certificate
+import java.util.concurrent.ConcurrentHashMap
+
+@Component(service = [LocallyHostedIdentitiesService::class])
+class LocallyHostedIdentitiesServiceImpl(
+    coordinatorFactory: LifecycleCoordinatorFactory,
+    private val subscriptionFactory: SubscriptionFactory,
+    private val configurationReadService: ConfigurationReadService,
+    private val certificateFactory: CertificateFactory,
+    private val sleeper: ((Long) -> Unit),
```

One thing this made me think about is whether this delay could be causing issues due to blocking processing from the message bus (e.g. heartbeats expiring causing Kafka rebalances etc.). Having a very quick look at the RPC messaging pattern, it seems like processing of RPC requests is done asynchronously and doesn't block the polling thread, but worth thinking about it a bit more and see if there are issues that could be caused from this and whether we could avoid it.

I guess an obvious alternative could be let it fail and have the user retry, causing a slightly worse UX.

Btw for context, we also have some work in the backlog to start moving away from excess use of the RPC pattern. So, when we do that we could maybe stash registration requests in a compacted topic and process them in the background (where this wouldn't be an issue). @charlieR3

---

actually, on second look. The processing of `onNext` seems to be synchronous, it just processes the results of the futures asynchronously. So, if we are ending up doing the waiting within the `onNext` which I think is the case, it could be an issue.

---

We have a few options:
1. Remove the retry mechanism altogether. In all my tests so far, the locally hosted identity is already on the map at the point we are creating the member context.
2. Reduce the duration or amount of retries. (4 X 0.1 seconds?)
3. Replace the retry with a condition notification mechanism with a short timeout. This will make the method return as soon as we have the information.
4. Keep it as is. We have a few places when we send a command to another worker while handling a command. 
5. Change the network setup API. After publishing the identity, we can ask one of the membership workers if it has received it.
6. Change the RPC producer to be more asynchronous. 

My preference would be to go with 2.

---

Thanks for the detailed list of options!

Indeed, if the data is expected to be there very quickly (as you seem to be observing), option 2 sounds reasonable. We delay only for a little time (to avoid blocking the processing pipeline significantly / causing rebalances etc.) and if there are cases where data is delayed more, we eventually fail with a good error message asking the user to retry a bit later.